src/Catalog/Helpers/CatalogWriterHelper.cs

```diff
@@ -0,0 +1,164 @@
+
+            if (storage == null)
+            {
+                throw new ArgumentNullException(nameof(storage));
+            }
+
+            if (maxDegreeOfParallelism < 1)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(maxDegreeOfParallelism),
+                    string.Format(Strings.ArgumentOutOfRange, 1, int.MaxValue));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var writer = new AppendOnlyCatalogWriter(storage, telemetryService, Constants.MaxPageSize);
+
+            var lastDate = DetermineLastDate(lastCreated, lastEdited, createdPackages);
+
+            if (packages.Count == 0)
+            {
+                return lastDate;
+            }
+
+            // Flatten the sorted list.
+            var workItems = packages.SelectMany(
+                    pair => pair.Value.Select(
+                        details => new PackageWorkItem(pair.Key, details)))
+                .ToArray();
+
+            await workItems.ForEachAsync(maxDegreeOfParallelism, async workItem =>
+            {
+                workItem.PackageCatalogItem = await packageCatalogItemCreator.CreateAsync(
+                    workItem.FeedPackageDetails,
+                    workItem.Timestamp,
+                    cancellationToken);
+            });
+
+            lastDate = packages.Last().Key;
+
+            // AppendOnlyCatalogWriter.Add(...) is not thread-safe, so add them all at once on one thread.
+            foreach (var workItem in workItems.Where(workItem => workItem.PackageCatalogItem != null))
+            {
+                writer.Add(workItem.PackageCatalogItem);
+
+                logger?.LogInformation("Add metadata from: {PackageDetailsContentUri}", workItem.FeedPackageDetails.ContentUri);
+            }
+
+            if (createdPackages.HasValue)
+            {
+                lastEdited = !createdPackages.Value ? lastDate : lastEdited;
+
+                if (updateCreatedFromEdited)
+                {
+                    lastCreated = lastEdited;
+                }
+                else
+                {
+                    lastCreated = createdPackages.Value ? lastDate : lastCreated;
+                }
+            }
+
+            var commitMetadata = PackageCatalog.CreateCommitMetadata(writer.RootUri, new CommitMetadata(lastCreated, lastEdited, lastDeleted));
+
+            await writer.Commit(commitMetadata, cancellationToken);
+
+            logger?.LogInformation("COMMIT metadata to catalog.");
+
+            return lastDate;
+        }
+
+        private static DateTime DetermineLastDate(DateTime lastCreated, DateTime lastEdited, bool? createdPackages)
```

Comment:
It does not matter which of the lastEdited lastCreated is newer? If there are certain expectations for the order should add logs if the expectations are not met?

Reply:
This is existing code which just moved to another type.